python-for-android/dists/kolibri/src/main/java/org/learningequality/task/Sentinel.java

```diff
@@ -0,0 +1,237 @@
+                .where(jobStatus)
+                .orderBy(JobStorage.Jobs.time_updated, false);
+    }
+
+    private WorkQuery buildWorkQuery(Bundle result) {
+        String requestId = JobStorage.Jobs.worker_extra.getValue(result);
+
+        if (requestId == null) {
+            String id = JobStorage.Jobs.id.getValue(result);
+            Log.v(TAG, "No request ID found for job " + id);
+            return Builder.TaskQuery.from(id).build();
+        }
+
+        return Builder.TaskQuery.from(UUID.fromString(requestId)).build();
+    }
+
+    /**
+     * Check for jobs with the given status and reconcile them with WorkManager
+     * Defaults to flagging missing work in WorkManager
+     *
+     * @param stateRef The job status in the Kolibri database for which to find jobs
+     * @return A future that will complete when all jobs have been checked, with a list of jobs
+     */
+    public CompletableFuture<Result[]> check(StateMap stateRef) {
+        return check(false, stateRef);
+    }
+
+    /**
+     * Check for jobs with the given status and reconcile them with WorkManager
+     * @param ignoreMissing Whether to ignore missing work in WorkManager
+     * @param stateRef The job status in the Kolibri database for which to find jobs
+     * @return A future that will complete when all jobs have been checked, with a list of jobs
+     */
+    public CompletableFuture<Result[]> check(
+            boolean ignoreMissing,
+            StateMap stateRef
+    ) {
+        Log.d(TAG, "Checking for jobs in state " + stateRef.getJobState());
+        SelectQuery query = buildQuery(stateRef.getJobState());
+        Bundle[] jobs = query.execute(db);
+
+        if (jobs == null || jobs.length == 0) {
+            Log.v(TAG, "No jobs to reconcile for status " + stateRef);
+            return CompletableFuture.completedFuture(null);
+        }
+
+        Log.d(TAG, "Cross-referencing " + jobs.length + " jobs with work manager");
+        return check(jobs, ignoreMissing, stateRef.getWorkInfoStates());
+    }
+
+    /**
+     * Check for the given jobs (Bundles) and reconciles them with WorkManager
+     *
+     * @param jobs The jobs to check
+     * @param ignoreMissing Whether to ignore missing work in WorkManager
+     * @param expectedWorkStates The expected WorkManager states for the found jobs
+     * @return A future that will complete when all jobs have been checked, with a list of jobs
+     */
+    public CompletableFuture<Result[]> check(
+            Bundle[] jobs,
+            boolean ignoreMissing,
+            WorkInfo.State... expectedWorkStates
+    ) {
+        final CompletableFuture<Result[]> future = new CompletableFuture<>();
+        final List<Result> allResults = new ArrayList<Result>(jobs.length);
+        CompletableFuture<List<Result>> chain = CompletableFuture.completedFuture(allResults);
+
+        for (Bundle job : jobs) {
+            chain = chain.thenComposeAsync((results) -> {
```

Comment:
So, am I reading it correctly that now each job is being processed sequentially here, whereas previously there could be parallel processing of individual jobs?

Reply:
Yep! The `thenComposeAsync` accepts a callback that itself returns another future, sort of like promise chaining, making everything process sequentially. This particular change seemed to provide the most improvement, and it makes the logs easier to read too. 

My only worry was how we managed the DB connection without a try-with resources statement, but rather rely on the future completing in order to close the connection. CoPilot indicated there could be issues mixing futures and the try-with. I was also happy that CoPilot had nothing to say other than 'no issues could be detected and the code is well structured [....]'. With the addition of some timeouts, it should also prevent any stalling should it re-occur.